src/main/kotlin/com/hibob/lambdaWithReceiver/lambdaWithReceiver.kt

```diff
@@ -0,0 +1,117 @@
+3. Implement pretty print using buildString function
+4. Make SpidermanMovieProduceActions nullable (if not so yet) and make the relevant adjustments
+*/
+
+fun main() {
+    val movie = createGoodMovie()
+    val runner = Runner(movie)
+    val success = runner.init()
+
+    printSuccessMessage(success)
+    println("Pretty print: ${movie.prettyPrint()}")
+//    println("Json: ${movie.toJson()}")
+}
+
+fun createGoodMovie() : SpidermanNoWayHome{
+    return SpidermanNoWayHome()
+}
+
+fun printSuccessMessage(success: Boolean) {
+    if (success) {
+        println("The movie was successful")
+    } else {
+        println("The movie failed")
+    }
+}
+
+interface SpidermanMovieProduceActions {
+    fun signTobeyMaguire()
+    fun signAndrew()
+    fun signTom()
+    fun getVillains()
+    fun isThereLockdown(): Boolean
+    fun publish():Boolean
+
+    val title: String
+    val airDate: LocalDate
+    val imdbRank: Double
+}
+
+
+class SpidermanNoWayHome() : SpidermanMovieProduceActions {
+
+    override val title: String = "Spiderman - No Way Home"
+    override val airDate: LocalDate = LocalDate.of(2021,12,16)
+    override val imdbRank: Double = 9.6
+
+    fun prettyPrint(): String = buildString {
+        appendLine("Title $title")
+        appendLine("Air date $airDate")
+        appendLine("MDB rank $imdbRank")
+    }
+
+    override fun signTobeyMaguire() {
+        //  Tobey signed!
+    }
+
+    override fun signAndrew() {
+        //    Andrew signed
+    }
+
+    override fun signTom() {
+        //    Tom signed
+    }
+
+    override fun getVillains() {
+        //   Got villains
+    }
+
+    override fun isThereLockdown(): Boolean = false
+
+    override fun publish(): Boolean = true
+
+//    fun toJson(): JsonNode {
+//        /* implement the following json structure:
+//                {
+//                 "title" : title,
+//                 "airDate": 2021-12-16,
+//                 "imdbRank": 9.6
+//                }
+//
+//        Note: In kotlin we receive the default object serializer "for free"
+//        and we will not have to implement it from here
+//        But, knowing how to write jsons in kotlin is still very important!
+//        the common use cases: S2S clients implementation, tests and more */
+//
+//        return JsonBuilderObject().json {
+//            "notImplemented" - true
+//        }.asTree()
+//    }
+}
+
+fun buildString(actions: StringBuilder.() -> Unit):String{
+    val builder = StringBuilder()
+    builder.actions()
+    return builder.toString()
+}
+
+class Runner(private val movieProducer: SpidermanMovieProduceActions?){
+    fun init(): Boolean {
+        return movieProducer?.run {
+            if (!isThereLockdown()) {
+                signTobeyMaguire()
+                signAndrew()
+                signTom()
+                getVillains()
+                publish()
+            } else false
+        } ?: return false
```

Comment:
just note that since you `return movieProducer?.run`  you can simply return false at the end instead of using ?:
 or keep it your way with the the extra `return` before the false - you already return at line 107

Reply:
I'm not sure I understand. 
Do you mean something like this?

fun init(): Boolean {
        if (movieProducer == null) return false
        
        return if (!movieProducer.isThereLockdown()) {
            movieProducer.signTobeyMaguire()
            movieProducer.signAndrew()
            movieProducer.signTom()
            movieProducer.getVillains()
            movieProducer.publish()
            true
        } else {
            false
        }
    }
    
    Because if I write it like this then i'm not implementing the function with lambda with a receiver as taught in the session. Unless I don't understand what you meant

Reply:
You code it ok now without the redundant return  - just another way of writting

class Runner(private val movieProducer: SpidermanMovieProduceActions?){
    fun init(): Boolean {
       movieProducer?.run {
            return if (!isThereLockdown()) {
                signTobeyMaguire()
                signAndrew()
                signTom()
                getVillains()
                publish()
            } else false
        } 
        return false
}

Reply:
No need to change

Reply:
ok got it, thanks! :)